### ISSUE TITLE
Kerberos support on LDAP cleartext

### ISSUE THREAD

**Issue by CravateRouge** [cannatag/ldap3#1026]
It would be great to be able to use Kerberos without LDAPS. For now, I've the following error when I'm trying to do so:
```
File "venv\lib\site-packages\ldap3\core\connection.py", line 615, in bind
    response = self.do_sasl_bind(controls)
  File "venv\lib\site-packages\ldap3\core\connection.py", line 1343, in do_sasl_bind
    result = sasl_gssapi(self, controls)
  File "venv\lib\site-packages\ldap3\protocol\sasl\kerberos.py", line 111, in sasl_gssapi
    return _windows_sasl_gssapi(connection, controls)
  File "venv\lib\site-packages\ldap3\protocol\sasl\kerberos.py", line 279, in _windows_sasl_gssapi
    status = winkerberos.authGSSClientStep(ctx, base64.b64encode(in_token).decode('utf-8'),
TypeError: Expected a channel bindings object
```
I suspect it is due to winkerberos using only the TGT even when a service ticket is already available for the ldap service.

The impacket kerberos module doesn't have this limitation and is working on both linux and windows and supports keytabs. Maybe it would be interesting to base ldap3 on it or use some of the code if this limitation is due to Ldap3 and not the kerberos modules.

**Comment by zorn96** [comment]
hi @CravateRouge ! based on the fact that you're on windows, I'm going to assume you're using AD.

ldap3 does support kerberos _authentication_ on cleartext (krb5). but it doesn't currently support LDAP signing/sealing layers (effectively krb5i and krb5p). Active Directory has 2 policies around requiring signing and requiring channel bind tokens.

signing requirements can be satisfied by TLS, and channel bind token can (I believe) be satisfied by sealing since a kerberos service tickets are inherently bound to a protocol. Since ldap3 can't do signing/sealing with kerberos, you need to use TLS to satisfy that AD policy. so that's why this happens (and you could alleviate it by getting your clients set to `NEGOTIATE` rather than `REQUIRED` for such policies).

regarding how to enhance things here. I added the `winkerberos` support for kerberos auth with ldap3 on windows, but my goal in doing so was just to bring windows support up to the same level as posix support, since those could use `gssapi`.
I don't think impacket should be used, because their libraries aren't very modular, and the library that would include their kerberos stuff has a lot of components that would throw up red flags in a lot of secure environments due their suitability for scanning/probing in malicious tools. but if the license of impacket allows it, then we could potentially copy some of their implementation over with the appropriate credit given back

**Comment by CravateRouge** [comment]
Hi @zorn96 , thanks for quickly answering! I omit to say I'm working on AD, you're right. More precisely on a 2019 DC with a win 2016 functional level.

That's what I understood and I only wanted to attempt a kerberos authentication but it seems that `winkerberos` doesn't even try the authentication when I'm not using LDAPS. See the following network capture corresponding to the `TypeError: Expected a channel bindings object` message:
[krbLdapWinkerberos.pcapng.gz](https://github.com/cannatag/ldap3/files/8237450/krbLdapWinkerberos.pcapng.gz)
So it doesn't seem to be linked to AD policy because you can clearly see that there is no error message from the DC when Ldap3 try to connect to the DC.
Here is the Ldap3 call I'm doing if you need more insight:
```
s = ldap3.Server("ldap://WIN-IJ5B521UO5L.bloody.local", get_info=ldap3.ALL)
c = ldap3.Connection(s, authentication=ldap3.SASL, sasl_mechanism=ldap3.KERBEROS, raise_exceptions=True)
c.bind()
```

However, when I'm using `GetADUsers.py` from impacket, I can clearly see a sasl bind (I don't know if the NTLMSSP negotiation before is also linked):
[krbLdapImpacket.pcapng.gz](https://github.com/cannatag/ldap3/files/8237448/krbLdapImpacket.pcapng.gz)

And I'm using the following tickets (ldap3 use only the TGT and impacket the ST):
```
[00000000] - 0x00000012 - aes256_hmac
   Start/End/MaxRenew: 12/03/2022 10:44:42 ; 12/03/2022 20:44:42 ; 13/03/2022 10:44:43
   Server Name       : krbtgt/BLOODY @ BLOODY.LOCAL
   Client Name       : Administrator @ BLOODY.LOCAL
   Flags a1c20000    : ? ; initial ; renewable ; invalid ; forwarded ; reserved ;

[00000001] - 0x00000012 - aes256_hmac
   Start/End/MaxRenew: 12/03/2022 10:50:54 ; 12/03/2022 20:44:42 ; 01/01/1601 01:00:00
   Server Name       : ldap/WIN-IJ5B521UO5L.bloody.local @ BLOODY.LOCAL
   Client Name       : Administrator @ BLOODY.LOCAL
   Flags 00250000    : name_canonicalize ; ok_as_delegate ; pre_authent ;
```

**Comment by CravateRouge** [comment]
After some digging I saw what's wrong: https://github.com/cannatag/ldap3/blob/040e09fe37cf06801960d494dfadf0a6ed94e38f/ldap3/protocol/sasl/kerberos.py#L279-L280
If `get_channel_bindings` return `None` and it will if I don't use LDAPS, `authGSSClientStep` will return an error. But if `channel_bindings` is not specified no error will be returned and it will run smoothly with LDAP cleartext.

However, I opened an issue regarding the use of only a service ticket https://github.com/mongodb/winkerberos/issues/48#issue-1167270253

**Comment by CravateRouge** [comment]
I saw another issue, regarding:
https://github.com/cannatag/ldap3/blob/040e09fe37cf06801960d494dfadf0a6ed94e38f/ldap3/protocol/sasl/kerberos.py#L269

`target_name` in my case will be `ldap@WIN-IJ5B521UO5L.bloody.local` but by default my DC doesn't have a matching SPN. However, `ldap/WIN-IJ5B521UO5L@BLOODY.LOCAL` is a valid one. You can see https://github.com/mongodb/winkerberos/issues/48#issuecomment-1067698843 for more details but I think https://github.com/cannatag/ldap3/blob/040e09fe37cf06801960d494dfadf0a6ed94e38f/ldap3/protocol/sasl/kerberos.py#L114 should be modified to match a default SPN.

**Comment by CravateRouge** [comment]
The issues are all solved with winkerberos 0.9.0
